washingtonpost/FastFEC
# feat: refactor python bindings, add line callback

This is a significant refactor of the Python bindings to be more library-driven, with the purpose of pushing a Python package to the Python package index (PyPI) as soon as possible. In the process, FastFEC was modified to provide custom line callback functionality for the sake of providing a convenient Python API that mimics other popular/inspirational packages such as [fecfile](https://github.com/esonderegger/fecfile).

The main file that drives this PR is `python/fastfec.py`. See this file for more detailed comments on how the API works. The top-level `setup.py` file is provided as a proof-of-concept that the Python package can be built automatically (not as a necessarily viable end result yet).

## Test Steps

If you haven't already, ensure you can build the fastfec library by running `zig build` (the steps in the README outline how to do this, but tl;dr `brew install zig --head` if you don't yet have `zig` installed). This will ensure the latest shared library file is discoverable by Python (which will allow all the new changes to work).

You can test the library by running a Python REPL (`python`) in the root directory of the repo. Also, make sure you have a .fec file in this directory for testing (in the following commands, it is assumed you have `13360.fec` downloaded from https://docquery.fec.gov/dcdev/posted/13360.fec in this root directory — but you can sub in whatever .fec file you have handy or want to test).

### To test the line by line parsing, run:

```python
from python.fastfec import FastFEC

with open('13360.fec', 'rb') as f:
    with FastFEC() as fastfec:
        for form, line in fastfec.parse(f):
            print("GOT", form, line)
```

This should print line information for each line in the passed in .fec file.

### To test the file to output file parsing, run:

```python
from python.fastfec import FastFEC

with open('13360.fec', 'rb') as f:
    with FastFEC() as fastfec:
        fastfec.parse_as_files(f,'python_output')
```

This should return `1` to indicate success and output .csv files in the `python_output` directory corresponding to a successful parse.

### To test the file to output file custom parsing, run:

```python
import os
from pathlib import Path
from python.fastfec import FastFEC

# Custom open method
def open_output_file(filename, *args, **kwargs):
    filename = os.path.join('custom_python_output', filename)
    output_file = Path(filename)
    output_file.parent.mkdir(exist_ok=True, parents=True)
    return open(filename, *args, **kwargs)

with open('13360.fec', 'rb') as f:
    with FastFEC() as fastfec:
        fastfec.parse_as_files_custom(f, open_output_file)
```

This should return `1` to indicate success and output .csv files in the `custom_python_output` directory corresponding to a successful parse.
